[TASK] Improve ItemsProcFunc documentation (#389)
TYPO3-Documentation/TYPO3CMS-Reference-TCA: Documentation/ColumnsConfig/CommonProperties/ItemsProcFunc.rst

   :type: string (class->method reference)
   :Scope: Display / Proc.
   :Types: :ref:`check <columns-check>`, :ref:`select <columns-select>`, :ref:`radio <columns-radio>`

   PHP method which is called to fill or manipulate the items array.

>    VENDOR\\Extension\\UserFunction\\YourClass->yourMethod

Reviewer: In general it is recommended to use the actual FQCN with `::class` instead and then concatenate the method:

```php
\VENDOR\Extension\UserFunction\YourClass::class . '->yourMethod'
```

Big advantage: with an IDE or similar places like these are also updated on refactoring.

   The provided method will have an array of parameters passed to it. The items array is passed by reference
   in the key :php:`items`. By modifying the array of items, you alter the list of items. A method may throw an
   exception which will be displayed as a proper error message to the user.

   Passed parameters
     * :php:`items` (passed by reference)
     * :php:`config` (TCA config of the field)

>      * :php:`TSconfig`

Reviewer: ```suggestion
     * :php:`TSconfig` (currently active page TSconfig)
```

Author: Actually it's not the page tsconfig but this: https://docs.typo3.org/m/typo3/reference-tsconfig/master/en-us/PageTsconfig/TceForm.html?highlight=itemsprocfunc#itemsprocfunc

Do you know how to interlink to this?
It starts with `<t3tsconfig:` but I don't know what comes after. Any idea?

Reviewer: You can actually see it in your link: `itemsprocfunc`. :-)

All IDs/anchors are actually used for linking.

See https://docs.typo3.org/m/typo3/docs-how-to-document/master/en-us/WritingReST/Hyperlinks.html#from-one-manual-to-another

Author: Thank you. It seems most headers have a reference to them like `.. _itemsProcFunc:`. This is missing for this exact option.

     * :php:`table` (current table)
     * :php:`row` (current database record)
     * :php:`field` (current field name)

     .. versionadded:: 7.6

>      This property only exists if the field has a :ref:`flex parent <columns-flex>`.

Reviewer: ```suggestion
        The following parameter only exists if the field has a :ref:`flex parent <columns-flex>`.
```

     * :php:`flexParentDatabaseRow`

     .. versionadded:: 11.2

>      These properties are filled if the current record has an :ref:`inline parent <columns-inline>`.

Reviewer: ```suggestion
        The following parameters only exist if the current record has an :ref:`inline parent <columns-inline>`.
```

     * :php:`inlineParentUid`
     * :php:`inlineParentTableName`
     * :php:`inlineParentFieldName`
     * :php:`inlineParentConfig`
     * :php:`inlineTopMostParentUid`
     * :php:`inlineTopMostParentTableName`
     * :php:`inlineTopMostParentFieldName`

   Example

>      The configuration for the field `category_field` of the table `tt_content` looks like this:

Reviewer: ```suggestion
     The configuration for a custom field :sql:`category_field` within the table :sql:`tt_content` could look like this:
```

     .. code-block:: php

        'type' => 'select',
        'renderType' => 'selectSingle',

>         'itemsProcFunc' => 'TYPO3\CMS\Core\Category\CategoryRegistry->getCategoryFieldsForTable',

Reviewer: ```suggestion
        'itemsProcFunc' => \TYPO3\CMS\Core\Category\CategoryRegistry::class . '->getCategoryFieldsForTable',
```

        'itemsProcConfig' => [
           'table' => 'tt_content'
        ],
        'minitems' => 0,
        'maxitems' => 1,
        'size' => 1

     The referenced :php:`itemsProcFunc` should populate the items by filling :php:`$configuration['items']`:

     .. code-block:: php

        /**
         * Returns a list of category fields for a given table for populating selector "category_field"
         * in tt_content table (called as itemsProcFunc).
         *
         * @param array $configuration Current field configuration
         * @throws \UnexpectedValueException

>          * @internal

Reviewer: ```suggestion
```

         */
        public function getCategoryFieldsForTable(array &$configuration)
        {
           $table = $configuration['config']['itemsProcConfig']['table'] ?? '';
           // Return early if no table is defined
           if (empty($table)) {
              throw new \UnexpectedValueException('No table is given.', 1381823570);
           }
           // Loop on all registries and find entries for the correct table
           foreach ($this->registry as $tableName => $fields) {
              if ($tableName === $table) {
                 foreach ($fields as $fieldName => $options) {
                    $fieldLabel = $this->getLanguageService()->sL($GLOBALS['TCA'][$tableName]['columns'][$fieldName]['label']);
                    $configuration['items'][] = [$fieldLabel, $fieldName];
                 }
              }
           }
        }

     .. note::

>         This example uses a non-existing TCA property :php:`itemsProcConfig` in order to pass an own argument

Reviewer: ```suggestion
        This example uses a custom TCA option :php:`itemsProcConfig` in order to pass an own argument
```
